2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,383 @@
+
+Current design of PowerShellGet is to have different cmdlets for each of the
+different resources types that it supports:
+
+`Find-Command`
+`Find-DscResource`
+`Find-Module`
+`Find-RoleCapability`
+`Find-Script`
+
+Instead, these will be abstracted by a `Find-PSResource` cmdlet.
+A `-Type` parameter will accept an array of types to allow filtering.
+
+With support of the generic `PSResource`, this means we can also find and
+install arbitrary nupkgs that may only contain an assembly the user wants to
+use in their script.
+
+For each repository, if the local cache does not exist, the cmdlet will call
+`Update-PSResourceCache` first and then attempt a search.
+However, this cmdlet does not update the cache if one exists.
+
+`Find-Module` will be retained to provide compatibility with v2.
+
+### Installing resources
+
+`Install-PSResource` will only work for modules unless `-DestinationPath` is
+specified which works for all resource types.
+A `-Prerelease` switch allows installing prerelease versions.
+Other types will use `Save-PSResource` (see below).
+A `-Repository` parameter accepts a specific repository name or URL to the repository:
+
+```powershell
+Install-PSResource myModule -Repository 'https://mygallery.com'
+
+# or
+
+Install-Module myModule -Repository 'https://mygallery.com'
+```
+
+If `-Repository` is not specified and the same resource is found in multiple
+repositories, then the resource is automatically installed quietly from the
+trusted repository with the highest version matching the `-Version` parameter
+(if specified, otherwise newest non-prerelease version unless `-Prerelease`
+is used).
+If there are no trusted repositories matching the query, then the newest version
+fulfilling the query will be prompted to be installed.
+If there are multiple repositories with the same trust level containing the same
+version, the first one is used.
+
+`-TrustRepository` can be used to suppress being prompted for untrusted sources.
+`-IgnoreDifferentPublisher` can be used to suppress being prompted if the publisher
+of the module is different from the currently installed version.
+`-Reinstall` can be used to re-install a module.
+
+A `-DestinationPath` parameter allows specifying the target directory instead
+of the default one.
+This will be in a different parameter set than `-Scope`.
+
+A `-NoClobber` switch will prevent installing modules that have the same cmdlets
+as a differently named module already on the system.
+
+A `-Quiet` switch will suppress progress information.
+
+`-Scope` with `AllUsers` and `CurrentUser` will work the same as it does in v2.
+Default is `CurrentUser`.
+
+`Install-Module` cmdlet will be retained for compatibility with v2.
+
+A change from v2 is that installed modules use the full semver (if used) of
+the module, so PSReadLine would be in a folder called `2.0.0-beta4` instead of
+just `2.0.0` which resolves issues of attempting to overwrite assemblies currently
+loaded on Windows.
```

Comment:
This would be awesome, but I have not seen the RFC for switching module versions to SemVer ...

Doing this would require you to handle SemVer for module loading, at least -- since we'd have to be able to tell that `2.0.0` is newer than `2.0.0-beta`. Will we be getting SemVer support in module manifests then? Will we be able to specify that in dependencies, so we can keep taking dependencies on betas? Right now if I require PSReadLine 2.0.0, the beta satisfies that dependency, but if you handled SemVer -- it would not!

Also, if I have a module `NeedsToWork` that has a dependency on `CoolKidsModule` with ModuleVersion = 1.0.0 (i.e. the poorly named _minimum_ version) and MaximumVersion 2.0  ... and I've installed CoolKidsModule version 1.3 .... 

When the cool kids release a 1.5.0-beta, if I install that side-by-side ... is it going to get used accidentally by `NeedsToWork` even though that module didn't (and shouldn't) opt into pre-releases?

Reply:
Marked https://github.com/PowerShell/PowerShell/issues/2983 for PS7

Reply:
Realistically, we'll have to wait for .NET 5 to have a semver type we can rely on.  For now, I'm going to change this language to use the current design which is non-semver folder.